src/query/functions/aggregation/base.go

```diff
@@ -0,0 +1,144 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package aggregation
+
+import (
+	"fmt"
+
+	"github.com/m3db/m3/src/query/block"
+	"github.com/m3db/m3/src/query/executor/transform"
+	"github.com/m3db/m3/src/query/functions/utils"
+	"github.com/m3db/m3/src/query/parser"
+)
+
+type aggregationFn func(values []float64, indices []int) float64
+
+var aggregationFunctions = map[string]aggregationFn{
+	SumType:               sumFn,
+	MinType:               minFn,
+	MaxType:               maxFn,
+	AverageType:           averageFn,
+	StandardDeviationType: stddevFn,
+	StandardVarianceType:  varianceFn,
+	CountType:             countFn,
+}
+
+// NodeParams contains additional parameters required for aggregation ops
+type NodeParams struct {
+	// MatchingTags is the set of tags by which the aggregation groups output series
+	MatchingTags []string
+	// Without indicates if series should use only the MatchingTags or if MatchingTags
+	// should be excluded from grouping
+	Without bool
+}
+
+// NewAggregationOp creates a new aggregation operation
+func NewAggregationOp(
+	opType string,
+	params NodeParams,
+) (parser.Params, error) {
+	if fn, ok := aggregationFunctions[opType]; ok {
+		return newBaseOp(params, opType, fn), nil
+	}
+	return baseOp{}, fmt.Errorf("operator not supported: %s", opType)
+}
+
+// baseOp stores required properties for count
+type baseOp struct {
+	params NodeParams
+	opType string
+	aggFn  aggregationFn
+}
+
+// OpType for the operator
+func (o baseOp) OpType() string {
+	return o.opType
+}
+
+// String representation
+func (o baseOp) String() string {
+	return fmt.Sprintf("type: %s", o.OpType())
+}
+
+// Node creates an execution node
+func (o baseOp) Node(controller *transform.Controller, _ transform.Options) transform.OpNode {
+	return &baseNode{
+		op:         o,
+		controller: controller,
+	}
+}
+
+func newBaseOp(params NodeParams, opType string, aggFn aggregationFn) baseOp {
+	return baseOp{
+		params: params,
+		opType: opType,
+		aggFn:  aggFn,
+	}
+}
+
+type baseNode struct {
+	op         baseOp
+	controller *transform.Controller
+}
+
+// Process the block
+func (n *baseNode) Process(ID parser.NodeID, b block.Block) error {
+	stepIter, err := b.StepIter()
+	if err != nil {
+		return err
+	}
+
+	params := n.op.params
+	buckets, metas := utils.GroupSeries(
+		params.MatchingTags,
+		params.Without,
+		n.op.opType,
+		stepIter.SeriesMeta(),
+	)
+
+	builder, err := n.controller.BlockBuilder(stepIter.Meta(), metas)
```

Comment:
will stepIter.Meta() change ? What if the aggregation is over something which is common across all series ?

Reply:
Added flatten/dedupe steps here